# Homework 4

Neural network exercises for CS-344, Professor Keith Vander Linden, Calvin University.

Answers by Nathan Meyer (tnm6).

## 4.1

Speculate on whether so-called "deep" neural networks will be another bust or really are a breakthrough for the foreseeable future.

### Response

My belief is that neural networks are more of a breakthrough than a bust. While technology is destined to change, and so approaches to AI will surely evolve, machine learning and especially deep neural networks shift the paradigm so drastically. More importantly, they are conceptually scalable. This is because the extent to which these AIs can "learn" is conceptually infinite, like how bits in binary code can be built into infinitely complex computer systems. A super-intelligence AI could conceivably be built upon sub-models of neural networks dedicated to performing well in specific areas, extending the abilities that the AI can have. Since it is based on the way neurons in our brains learn, this is, at least abstractly, similar to how the human brain can learn complex processes.

Take the work that is being done at Google. Translation was considered to be impossible to perform reliably until Google Brain was thrown at it, where performance improved at rates previously unseen before, particularly with complex performance metrics like the "artistry" of the language found in poems, etc. Similar work has been done with image classification. Functions like these are specific in and of themselves, but stacked or combined together, they can, perceivably, reach some manner of flexibility in actions performed like the human brain can. The extent to which this could scale is conceptually infinite.

The problem is data. These concepts have been around for decades, but only now, with the advent of the internet, have there been datasets large enough to even begin to train neural networks to perform reliably in many areas (like translation). The limit, then, is not so much the conceptual limits of the AI design, but the resources required to "feed" the learning models. For now, that training data is highly dependent upon reliable human data entry. Further, training these networks are highly demanding upon hardware. However, while these two issues may limit the breadth of neural network AI for now, improvements in both of them in the future would correlate with greater performance in neural networks. Since hardware improvements are expected (for now), and streamlining or improving data entry may still be possible, I do not see why neural networks will not continue to be the way forward.

## 4.2

Hand-compute a single, complete back-propagation cycle. Use the example network from class and compute the updated weight values for the first gradient descent iteration for the XOR example, i.e., [1, 1] → 0. Use the same initial weights we used in the class example but assume the identity function as the activation function (f(x) = x).

### Computation

1. Fill in random weights (same as class example).

    $\begin{aligned}
    &\begin{bmatrix}
    w_{i_1,h_1} & w_{i_1,h_2} \\
    w_{i_2,h_1} & w_{i_2,h_2}
    \end{bmatrix}
    \leftarrow
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} \\
    &\begin{bmatrix}
    w_{h_1, o_1} \\ 
    w_{h_2, o_1} 
    \end{bmatrix}
    \leftarrow
    \begin{bmatrix}
    0.14 \\
    0.15
    \end{bmatrix}
    \end{aligned}$

2. Compute the output for sample (XOR: `[1, 1]` &rarr; `0`)

    $\begin{aligned}
    o_j &= 
    \begin{bmatrix}
    1 & 1 \\ 
    \end{bmatrix}
    \cdot
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix}
    \cdot
    \begin{bmatrix}
    0.14 \\
    0.15
    \end{bmatrix}
    \\ &=
    \begin{bmatrix}
    1 * 0.11 + 1 * 0.21 & 1 * 0.12 + 1 * 0.08
    \end{bmatrix}
    \cdot
    \begin{bmatrix}
    0.14 \\ 
    0.15
    \end{bmatrix}
    \\ &=
    \begin{bmatrix}
    0.32 & 0.20
    \end{bmatrix}
    \cdot
    \begin{bmatrix}
    0.14 \\ 
    0.15 
    \end{bmatrix}
    \\ &=
    \begin{bmatrix}
    0.32 * 0.14 + 0.20 * 0.15
    \end{bmatrix}
    \\ &= 0.0748
    \end{aligned}
    \\
    $

3. Compute the error and delta.

    $\begin{aligned}
    L_2Error &= (0 - 0.0748)^2 \\
    &= 0.0056 \\
    \Delta_{o_1} &= (0 - 0.0748) \\
    &= -0.0748 \\
    \end{aligned}$

4. Back-propagate through the network, assuming:
    $learning\_rate = 0.05$; 
    Activation function `f(x) = x`.
     
    $\begin{aligned}
    \begin{bmatrix}
    w_{h_1, o_1} \\ 
    w_{h_2, o_1}
    \end{bmatrix} &\leftarrow 
    \begin{bmatrix}
    0.14 \\ 
    0.15 
    \end{bmatrix} + 0.05 \cdot 
    \begin{bmatrix}
    0.32 \\ 
    0.20
    \end{bmatrix} \cdot 1.0 \cdot -0.0748 \\\\
    &= 
    \begin{bmatrix}
    0.14 \\ 
    0.15 
    \end{bmatrix} + 
    \begin{bmatrix}
    0.05 * 0.32 * 1.0 * -0.0748 \\
    0.05 * 0.20 * 1.0 * -0.0748 
    \end{bmatrix} \\
    &= 
    \begin{bmatrix}
    0.14 \\ 
    0.15 
    \end{bmatrix} +
    \begin{bmatrix}
    -0.001197 \\
    -0.000748
    \end{bmatrix} \\
    &=
    \begin{bmatrix}
    0.138803 \\ 
    0.149252
    \end{bmatrix}
    \end{aligned}$

    $\begin{aligned}
    \begin{bmatrix}
    w_{i_1,h_1} & w_{i_1,h_2} \\ 
    w_{i_2,h_1} & w_{i_2,h_2}
    \end{bmatrix} &\leftarrow 
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} + 0.05 \cdot
    \begin{bmatrix}
    1 & 1 \\ 
    1 & 1
    \end{bmatrix} \cdot 1.0 \odot
    \begin{bmatrix}
    0.14 & 0.15 \\ 
    0.14 & 0.15
    \end{bmatrix} \cdot -0.0748 \\ &=
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} + 
    \begin{bmatrix}
    0.05 * 1 * 1.0 & 0.05 * 1 * 1.0 \\ 
    0.05 * 1 * 1.0 & 0.05 * 1 * 1.0 \\ 
    \end{bmatrix} \odot 
    \begin{bmatrix}
    0.14 * -0.0748 & 0.15 * -0.0748 \\ 
    0.14 * -0.0748 & 0.15 * -0.0748 
    \end{bmatrix} \\ &=
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} + 
    \begin{bmatrix}
    0.05 & 0.05 \\ 
    0.05 & 0.05 
    \end{bmatrix} \odot 
    \begin{bmatrix}
    -0.010472 & -0.01122 \\
    -0.010472 & -0.01122
    \end{bmatrix} \\ &=
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} + 
    \begin{bmatrix}
    0.05 * -0.010472 & 0.05 * -0.01122 \\
    0.05 * -0.010472 & 0.05 * -0.01122
    \end{bmatrix} \\ &=
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} +     
    \begin{bmatrix}
    -0.000524 & -0.000561 \\
    -0.000524 & -0.000561
    \end{bmatrix} \\ &= 
    \begin{bmatrix}
    0.109476 & 0.119439 \\
    0.209476 & 0.079439
    \end{bmatrix}  
    \end{aligned}$
    

## 4.3

Build a Keras-based ConvNet for Keras’s Fashion MNIST dataset (fashion_mnist). Experiment with different network architectures, submit your most performant network, and report the results.

### Configuration

Prepare the data.

In [1]:
from keras.datasets import fashion_mnist
from keras.utils import to_categorical

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

Using TensorFlow backend.


Configure the CNN model. Conv/Relu-Conv/Relu-MaxPooling * 2.

In [3]:
from keras import layers
from keras import models

model = models.Sequential()

model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.Conv2D(32, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))

model.add(layers.Conv2D(64, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))

model.add(layers.Flatten())
model.add(layers.Dense(512, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

Compile and train the model.

In [6]:
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit(train_images, train_labels,
          epochs=10,
          batch_size=64,
          validation_data=(test_images, test_labels))

model.evaluate(test_images, test_labels)

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 79s 1ms/step - loss: 0.3439 - acc: 0.8748 - val_loss: 0.2901 - val_acc: 0.8953
Epoch 2/10
60000/60000 [==============================] - 80s 1ms/step - loss: 0.2566 - acc: 0.9063 - val_loss: 0.2788 - val_acc: 0.8947
Epoch 3/10
60000/60000 [==============================] - 87s 1ms/step - loss: 0.2188 - acc: 0.9195 - val_loss: 0.2981 - val_acc: 0.8933
Epoch 4/10
60000/60000 [==============================] - 83s 1ms/step - loss: 0.1904 - acc: 0.9307 - val_loss: 0.2565 - val_acc: 0.9118
Epoch 5/10
60000/60000 [==============================] - 82s 1ms/step - loss: 0.1725 - acc: 0.9382 - val_loss: 0.2480 - val_acc: 0.9138
Epoch 6/10
60000/60000 [==============================] - 87s 1ms/step - loss: 0.1557 - acc: 0.9435 - val_loss: 0.2799 - val_acc: 0.9162
Epoch 7/10
60000/60000 [==============================] - 84s 1ms/step - loss: 0.1434 - acc: 0.9481 - val_loss: 0.2563 - val_acc

[0.3433183187291026, 0.916]

### Results

Reconfigured network with Conv-Relu-Conv-Relu-Pool produced highest test accuracy I could attain, which was 91.6%. There might be simpler configurations which can achieve this accuracy more quickly, but I was unable to get results higher than that for the test set.